File: ios-uikit-lvt/LvtCrmUIKit/AvatarRepository.swift
import Foundation
import UIKit

extension Notification.Name {
    static let accountAvatarDidChange = Notification.Name("AccountAvatarDidChange")
}

let avatarMaxBytes = 2 * 1024 * 1024
let avatarMaxEdge: CGFloat = 512

actor AvatarRepository {
    private let convex: ConvexHttpClient
    private let tokenProvider: @Sendable () -> String?
    private let session: URLSession
    private var cachedKey: String?
    private var cachedImage: UIImage?

    init(convex: ConvexHttpClient, tokenProvider: @escaping @Sendable () -> String?) {
        self.convex = convex
        self.tokenProvider = tokenProvider
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = 45
        configuration.timeoutIntervalForResource = 60
        session = URLSession(configuration: configuration)
    }

    func image(userId: String, hasAvatar: Bool, version: String?) async -> UIImage? {
        guard hasAvatar else {
            cachedKey = nil
            cachedImage = nil
            return nil
        }
        let key = "\(userId):\(version ?? "current")"
        if cachedKey == key { return cachedImage }
        guard let token = tokenProvider(), !token.isEmpty,
              let url = URL(string: "\(ConvexConfig.webURL)/api/files/avatar") else { return cachedImage }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        do {
            let (data, response) = try await session.data(for: request)
            let status = (response as? HTTPURLResponse)?.statusCode ?? 0
            guard status == 200, !data.isEmpty, let image = UIImage(data: data) else { return cachedImage }
            cachedKey = key
            cachedImage = image
            return image
        } catch {
            return cachedImage
        }
    }

    func upload(image: UIImage) async throws -> UIImage {
        guard let payload = AvatarCrop.encodeWebP(image) else {
            throw ConvexException(code: "INVALID_AVATAR_FILE")
        }
        let stored = AvatarCrop.contentType(for: payload)
        let upload = try await convex.mutation("userAvatar:generateUploadUrl")
        guard let uploadUrlString = upload["value"] as? String,
              let uploadUrl = URL(string: uploadUrlString) else {
            throw ConvexException(code: "AVATAR_UPLOAD_FAILED")
        }
        var request = URLRequest(url: uploadUrl)
        request.httpMethod = "POST"
        request.setValue(stored.mime, forHTTPHeaderField: "Content-Type")
        request.httpBody = payload
        let (data, response) = try await session.data(for: request)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard status >= 200, status < 300,
              let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let storageId = json["storageId"] as? String, !storageId.isEmpty else {
            throw ConvexException(code: "AVATAR_UPLOAD_FAILED")
        }
        _ = try await convex.action("userAvatar:setOwnAvatar", args: [
            "storageId": storageId,
            "fileName": stored.fileName,
            "fileSize": payload.count,
        ])
        let saved = UIImage(data: payload) ?? image
        cachedKey = nil
        cachedImage = saved
        return saved
    }

    func clear() async throws {
        _ = try await convex.mutation("userAvatar:clearOwnAvatar")
        cachedKey = nil
        cachedImage = nil
    }
}

File: android-app/app/src/main/java/lvt/crm/data/auth/AvatarRepository.kt
package lvt.crm.data.auth

import android.graphics.Bitmap
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import lvt.crm.data.convex.ConvexConfig
import lvt.crm.data.convex.ConvexException
import lvt.crm.data.convex.ConvexHttpClient
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.File
import java.util.concurrent.TimeUnit

class AvatarRepository(
    private val convex: ConvexHttpClient,
    private val tokenProvider: () -> String?,
    cacheDir: File,
) {
    private val cacheRoot = File(cacheDir, "avatars").apply { mkdirs() }
    private val _bitmap = MutableStateFlow<Bitmap?>(null)
    val bitmap: StateFlow<Bitmap?> = _bitmap.asStateFlow()
    private val http = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(45, TimeUnit.SECONDS)
        .writeTimeout(45, TimeUnit.SECONDS)
        .build()

    suspend fun sync(userId: String, hasAvatar: Boolean, avatarVersion: String?) = withContext(Dispatchers.IO) {
        if (userId.isBlank() || !hasAvatar) {
            _bitmap.value = null
            return@withContext
        }
        val version = avatarVersion?.takeIf { it.isNotBlank() } ?: "current"
        val cached = File(cacheRoot, "${userId.hashCode()}-$version.jpg")
        if (cached.isFile && cached.length() > 0) {
            _bitmap.value = decodeAvatarBitmap(cached.readBytes())
            return@withContext
        }
        val token = tokenProvider()?.takeIf { it.isNotBlank() } ?: return@withContext
        val base = ConvexConfig.webUrl.trimEnd('/')
        val metadataRequest = Request.Builder()
            .url("$base/api/files/avatar/metadata")
            .header("Authorization", "Bearer $token")
            .get()
            .build()
        val metadata = http.newCall(metadataRequest).execute().use { response ->
            val body = response.body?.string().orEmpty()
            if (!response.isSuccessful) return@withContext
            runCatching { JSONObject(body) }.getOrNull()
        } ?: return@withContext
        val remoteVersion = metadata.optString("fileVersion").ifBlank { version }
        val versioned = File(cacheRoot, "${userId.hashCode()}-$remoteVersion.jpg")
        if (versioned.isFile && versioned.length() > 0) {
            _bitmap.value = decodeAvatarBitmap(versioned.readBytes())
            return@withContext
        }
        val request = Request.Builder()
            .url("$base/api/files/avatar")
            .header("Authorization", "Bearer $token")
            .get()
            .build()
        val bytes = http.newCall(request).execute().use { response ->
            if (!response.isSuccessful) return@withContext
            response.body?.bytes()
        } ?: return@withContext
        if (bytes.isEmpty()) return@withContext
        runCatching {
            versioned.parentFile?.mkdirs()
            versioned.writeBytes(bytes)
        }
        _bitmap.value = decodeAvatarBitmap(bytes)
    }

    suspend fun upload(bytes: ByteArray, fileName: String = "avatar.webp"): Result<Unit> =
        withContext(Dispatchers.IO) {
            runCatching {
                if (bytes.isEmpty()) throw ConvexException("INVALID_AVATAR_FILE", "INVALID_AVATAR_FILE")
                if (bytes.size > AvatarFile.MAX_BYTES) {
                    throw ConvexException("AVATAR_FILE_TOO_LARGE", "AVATAR_FILE_TOO_LARGE")
                }
                val contentType = when {
                    fileName.endsWith(".png", ignoreCase = true) -> "image/png"
                    fileName.endsWith(".jpg", ignoreCase = true) ||
                        fileName.endsWith(".jpeg", ignoreCase = true) -> "image/jpeg"
                    else -> "image/webp"
                }
                val uploadUrl = convex.mutation("userAvatar:generateUploadUrl").optString("value")
                if (uploadUrl.isBlank()) throw ConvexException("AVATAR_UPLOAD_FAILED", "AVATAR_UPLOAD_FAILED")
                val uploaded = http.newCall(
                    Request.Builder()
                        .url(uploadUrl)
                        .post(bytes.toRequestBody(contentType.toMediaType()))
                        .build(),
                ).execute().use { response ->
                    val body = response.body?.string().orEmpty()
                    if (!response.isSuccessful) {
                        throw ConvexException("AVATAR_UPLOAD_FAILED", "AVATAR_UPLOAD_FAILED")
                    }
                    JSONObject(body)
                }
                val storageId = uploaded.optString("storageId")
                if (storageId.isBlank()) throw ConvexException("AVATAR_UPLOAD_FAILED", "AVATAR_UPLOAD_FAILED")
                convex.action(
                    "userAvatar:setOwnAvatar",
                    JSONObject()
                        .put("storageId", storageId)
                        .put("fileName", fileName)
                        .put("fileSize", bytes.size),
                )
                _bitmap.value = decodeAvatarBitmap(bytes)
            }
        }

    suspend fun clear(): Result<Unit> = withContext(Dispatchers.IO) {
        runCatching {
            convex.mutation("userAvatar:clearOwnAvatar")
            _bitmap.value = null
        }
    }

    fun clearLocal() {
        _bitmap.value = null
    }
}
